migrations/2023-07-18-082614_post_aggregates_community_id/up.sql

```diff
@@ -0,0 +1,31 @@
+-- Your SQL goes here
+ALTER TABLE post_aggregates
+    ADD COLUMN community_id integer REFERENCES community ON UPDATE CASCADE ON DELETE CASCADE,
+    ADD COLUMN creator_id integer REFERENCES person ON UPDATE CASCADE ON DELETE CASCADE;
+
+CREATE OR REPLACE FUNCTION post_aggregates_post()
+    RETURNS trigger
+    LANGUAGE plpgsql
+AS
+$$
+BEGIN
+    IF (TG_OP = 'INSERT') THEN
+        INSERT INTO post_aggregates (post_id, published, newest_comment_time, newest_comment_time_necro, community_id,
+                                     creator_id)
+        VALUES (NEW.id, NEW.published, NEW.published, NEW.published, NEW.community_id, NEW.creator_id);
+    ELSIF (TG_OP = 'DELETE') THEN
+        DELETE FROM post_aggregates WHERE post_id = OLD.id;
+    END IF;
+    RETURN NULL;
+END
+$$;
+
+UPDATE post_aggregates
+SET community_id=post.community_id,
+    creator_id=post.creator_id
+FROM post
+WHERE post.id = post_aggregates.post_id;
+
+ALTER TABLE post_aggregates
+    ALTER COLUMN community_id SET NOT NULL,
+    ALTER COLUMN creator_id SET NOT NULL;
```

Comment:
After this one, much like the post table has indexes on community_id and creator_id, it would probably be a good idea to add these two indexes to `post_aggregates` now too:

IE:
 
```sql
create index idx_post_aggregates_community on post_aggregates (community_id);
create index idx_post_aggregates_creator on post_aggregates (creator_id);
```

Reply:
Sure, added these indexes as well

Reply:
Ok, the query plans are definitely consistently worse with those indexes.
With indexes: https://explain.depesz.com/s/W7cF
Without indexes: https://explain.depesz.com/s/4qds

Reply:
I reverted these indexes, let me know if you disagree with that

Reply:
Discussed a bit on Matrix already, but commenting here as well for visibility:

---

> the "right" index should be post_aggregates(community_id, featured_local desc, hot_rank desc)

This index also messes up the query plan. It indeed makes the queries super fast for users with few subscriptions, but it makes them slow again for others.

**Without that index**
User with 86 subscriptions: 3ms query execution
User with 0 subscriptions: 700 ms query execution

**With that index**
User with 86 subscrptions: 2-3 *second* query execution
User with 0 subscriptions: 1ms query execution

> phiresky:
> weird. do the query plans have larger misestimates?
> if so try setting default statistics target to 1000 and run analyze again. if not then I don't know why PG is being dumb 


--- 

I will see tomorrow if playing around with statistics will help. I know there's also CREATE STATISTICS` in postgres, but I've never used it so would need to investigate it a bit.

Reply:
The limit and sort is the main culprit, but also its weird to me that postgres sometimes chooses a slower index.

We can tweak those indexes later, I'll try to tag the issues as `DB`